server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1723,8 +1726,32 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
     /**
      * Returns a new {@link QueryRewriteContext} with the given {@code now} provider
      */
-    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis) {
-        return new QueryRewriteContext(parserConfig, client, nowInMillis);
+    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis, IndicesRequest indicesRequest) {
+        Supplier<Map<String, IndexMetadata>> indexMetadataMapSupplier = () -> {
+            // Skip unavailable indices
+            IndicesOptions.Builder indicesOptionsBuilder = IndicesOptions.builder(indicesRequest.indicesOptions())
+                .concreteTargetOptions(new IndicesOptions.ConcreteTargetOptions(true));
+            Index[] indices = indexNameExpressionResolver.concreteIndices(
```

Comment:
is there a chance to perhaps not resolve indices twice, but rather do it earlier in TransportSearchAction, so that the resolved indices can be provided here directly?

Reply:
Perhaps? I'm not too familiar with that part of the codebase. I'll take a look, but if you know of a spot where this would fit well, i'd love some pointers.

Reply:
I investigated a bit and have an idea of how to resolve indices no more than twice (at most once on the coordinating node, and once on the shard):

In [TransportSearchAction#executeRequest](https://github.com/elastic/elasticsearch/blob/249223fa430d4481440f73bd1dbca0e5959435cd/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L313), create a new index name(s) supplier that we pass to `searchService.getRewriteContext`. This can then be used to lazily build the index metadata map, similar to how it is done in this PR. This would all be executed on the coordinator node (as necessary).

The shard would then run [TransportSearchAction#executeLocalSearch](https://github.com/elastic/elasticsearch/blob/249223fa430d4481440f73bd1dbca0e5959435cd/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L917), which would resolve the local indices a (potential) second time. There may be a way to do this resolution once one the coordinator node and share it with the shard, but it's not immediately obvious to me and would be more likely to introduce some BwC breaking change.

WDYT? Am I on the right track?

Reply:
We discussed offline with @javanna and agreed that what you propose is an acceptable path. The main thing we need to avoid is to add computations for searches that don't contain a semantic query, that's what we're trying to optimise.

Reply:
I started implementing this change and hit an issue early on: `TransportSearchAction` has access to all the dependencies necessary to resolve the indices, but `SearchService#getRewriteContext` is also called from other places (ex: `TransportValidateQueryAction`, `TransportExplainAction`) that don't have access to the same dependencies. Even if they did, we would be duplicating the code to resolve the indices in each place `SearchService#getRewriteContext` is called, which doesn't seem like a good idea.

Based on this finding, it seems like a good idea to leave index resolution in `IndicesService` (as it is now), which contains all the necessary dependencies and prevents code duplication. It also maintains the key characteristic that index resolution is performed on the coordinator node only for semantic queries since this operation is performed lazily.

Thoughts?

Reply:
which dependencies are you referring to? I don't follow entirely the concerns around what happens in the coord node vs in the data node: my suggestion is if we do indices resolution in the coord node, which we must be doing,  could we do it earlier and share the result instead of doing it twice? Hopefully this impacts only the coordinating codepath.

Reply:
@javanna I'm referring to the dependencies `IndexNameExpressionResolver` & `ClusterService`. `TransportSearchAction` has those, but `TransportValidateQueryAction` & `TransportExplainAction` don't. `IndicesService` has both and is in the common codepath for all of the relevant `Transport` classes, so it seems like a good place to do this work.

> my suggestion is if we do indices resolution in the coord node, which we must be doing, could we do it earlier and share the result instead of doing it twice?

I'll research this a bit more, but in my previous pass at this I didn't find the codepath where the indices were resolved on the coordinator node. Maybe I misinterpreted the results though, I was testing on a single-node system.